<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 240%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 35px 15px; 
          border-radius: 25px 50px; 
          margin: 0;">
  Sales Forecasting for a Retail Chain
</p>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.7; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <h2 style="color:#0072ff; margin-top:0;">Problem Statement</h2>
  
  <p>
    This project focuses on <b>weekly sales forecasting</b> for <b>Walmart stores across the United States</b>, 
    leveraging historical data that captures key operational and economic factors influencing retail performance. 
    The dataset includes variables such as <b>temperature, fuel price, Consumer Price Index (CPI), unemployment rate,</b> 
    and <b>holiday indicators</b>, which play a crucial role in driving customer purchasing behavior and sales variations.
  </p>
  
  <p>
    The objective of this project is to develop a reliable <b>machine learning–based forecasting system</b> 
    using various <b>regression models</b> such as <b>Linear Regression, Ridge, Lasso, Random Forest,</b> 
    and <b>XGBoost</b>. 
    These models aim to predict future weekly sales accurately, uncover <b>trends and relationships</b> 
    between sales and external factors, and generate insights that support 
    <b>strategic decision-making</b> in <b>inventory management, staffing, and promotional planning</b> 
    for Walmart’s retail operations.
  </p>

</div>


<a id="contents_table"></a>

<div style="border-radius:18px; padding:24px; background-color:white; 
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.8; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

<h2 align="Left" style="color:#0072ff;"> Table of Contents</h2>

<ul style="list-style-type:none; padding-left:15px;">

<li><b><a href="#setup" style="color:#0072ff; text-decoration:none;">
Step 1 | Data Ingestion and Setup In SQL
</a></b>
  <ul>
    <li><a href="#libraries" style="color:#003366;">Step 1.1 | Importing Necessary Libraries</a></li>
    <li><a href="#load_dataset" style="color:#003366;">Step 1.2 | Generating Folder</a></li>
    <li><a href="#sql_creation" style="color:#003366;">Step 1.3 | Logging Configuration</a></li>
    <li><a href="#data_cleaning" style="color:#003366;">Step 1.4 | Database Ingestion Function</a></li>
    <li><a href="#script_execution" style="color:#003366;">Step 1.5 | Loading and Ingesting CSV files</a></li>
    <li><a href="#script_execution" style="color:#003366;">Step 1.6 | Script Excecution</a></li>
    <li><a href="#script_execution" style="color:#003366;">Step 1.7 | Data Loading and Info </a></li>
    <li><a href="#script_execution" style="color:#003366;">Step 1.8 | Extracting Relevent Columns for Analysis</a></li>



  </ul>
</li>

<br>

<li><b><a href="#eda" style="color:#0072ff; text-decoration:none;">
Step 2 | Exploratory Data Analysis (EDA)
</a></b>
  <ul>
    <li><a href="#feature_selection" style="color:#003366;">Step 2.1 | Time-Based Analysis</a></li>
    <li><a href="#store_economic" style="color:#003366;">Step 2.2 | Store-Level and Economic Insights</a></li>
  </ul>
</li>

<br>

<li><b><a href="#feature_engineering" style="color:#0072ff; text-decoration:none;">
Step 3 | Feature Engineering
</a></b>
  
  
</li>

<br>

<li><b><a href="#train_test" style="color:#0072ff; text-decoration:none;">
Step 4 | Train–Test Split and Baseline Setup
</a></b>
  <ul>
    
  </ul>
</li>

<br>

<li><b><a href="#model_training" style="color:#0072ff; text-decoration:none;">
Step 5 | Regression-Based Modeling
</a></b>
  <ul>
    <li><a href="#linear_regression" style="color:#003366;">Step 5.1 | Linear Regression Model</a></li>
    <li><a href="#regularized_models" style="color:#003366;">Step 5.2 | Ridge Model</a></li>
    <li><a href="#regularized_models" style="color:#003366;">Step 5.3 | Lasso Model</a></li>
    <li><a href="#ensemble_models" style="color:#003366;">Step 5.4 | Random Forest Model </a></li>
    <li><a href="#ensemble_models" style="color:#003366;">Step 5.5 | XGBoost Model</a></li>

  </ul>
</li>

<br>


<li><b><a href="#forecasting" style="color:#0072ff; text-decoration:none;">
Step 6 | Final Sales Forecasting and Future Predictions
</a></b>
</li>

<br>

<li><b><a href="#summary" style="color:#0072ff; text-decoration:none;">
Step 7 | Summary and Conclusion
</a></b>
  <ul>
    <li><a href="#findings" style="color:#003366;">Step 7.1 | Summary</a></li>
    <li><a href="#future_scope" style="color:#003366;">Step 7.2 | Bussiness Insights</a></li>
    <li><a href="#business_impact" style="color:#003366;">Step 7.3 | Conclusion</a></li>
  </ul>
</li>

</ul>
</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 1 | Data Ingestion and Setup In SQL
</p>


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.1 |</span><span style='color:#0072ff'> Importing Libraries</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.7; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <p>
    The script begins by importing the necessary <b>Python libraries</b>  
    and establishing a connection to the <b>database</b>.
  </p>

</div>


In [1]:
import pandas as pd               # For reading CSV files and handling dataframes
import os                         # For interacting with the operating system (folders, files)
from sqlalchemy import create_engine  # For connecting to SQLite database
import logging                    # For logging ingestion activity and errors
import time                       # For calculating ingestion time and generating timestamps
import plotly.io as pio
pio.renderers.default = "iframe_connected"


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.2 |</span><span style='color:#0072ff'> Folder</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:110%; font-family:Verdana, sans-serif; 
            line-height:1.6; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <p>
    This step ensures that essential folders exist before the ingestion process begins.  
    The script automatically creates two directories — one for <b>logs</b> and another for <b>data</b>.  
    This setup helps organize all input files and keep logs of every operation systematically.
  </p>

</div>


In [2]:
# Ensure required folders exist, create them if they don't
os.makedirs("logs", exist_ok=True)   # Folder for log files
os.makedirs("data", exist_ok=True)   # Folder to store CSV files to ingest


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.3 |</span><span style='color:#0072ff'>  Logging Configuration</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:110%; font-family:Verdana, sans-serif; 
            line-height:1.6; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <p>
    Logging helps track every stage of the data ingestion process.  
    The configuration below sets up a detailed logging system that records timestamps, 
    message levels, and process details into a dedicated <b>ingestion_db.log</b> file.  
    This ensures that any issue or success is properly documented for debugging or auditing.
  </p>

</div>


In [3]:
# Configure logging settings
logging.basicConfig(
    filename='logs/ingestion_db.log',           # Log file path
    level=logging.INFO,                         # Log all INFO and higher level messages
    format='%(asctime)s - %(levelname)s - %(message)s',  # Timestamp, level, and message
    filemode='a'                                # Append logs to existing file
)


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.4 |</span><span style='color:#0072ff'>  Database Ingestion Function</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:110%; font-family:Verdana, sans-serif; 
            line-height:1.6; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <p>
    This function handles inserting data from a Pandas DataFrame into a <b>SQLite database</b>.  
    It takes the table name and database connection as inputs and logs whether the operation succeeded.  
    The <b>mode</b> parameter controls whether to append to an existing table or overwrite it completely.
  </p>

</div>


In [4]:
def ingest_db(df, table_name, con, mode="append"):
    """
    Ingest the dataframe into a database table.

    Parameters:
    - df: pandas DataFrame to ingest
    - table_name: name of the table in SQLite
    - con: SQLAlchemy connection engine
    - mode: 'append' to add rows, 'replace' to overwrite table
    """
    df.to_sql(table_name, con=con, if_exists=mode, index=False)
    logging.info(f"Table '{table_name}' ingested successfully with mode='{mode}'.")


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.5 |</span><span style='color:#0072ff'>   Loading and Ingesting CSV Files</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:110%; font-family:Verdana, sans-serif; 
            line-height:1.6; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <p>
    This part automates the <b>data ingestion pipeline</b>.  
    It scans the <b>data</b> folder for all CSV files, reads each one into a DataFrame,  
    and loads it into the SQLite database.  
  </p>

  <p>
    The function also measures how long the ingestion takes and handles errors gracefully,  
    logging any issue with detailed messages.

</div>


In [5]:
import os
import time
import pandas as pd
import logging
import shutil
from sqlalchemy import create_engine

def ingest_db(df, table_name, con, mode="append"):
    """Helper function to ingest dataframe into SQLite database."""
    df.to_sql(table_name, con=con, if_exists=mode, index=False)

def load_raw_data(append_mode=True, kaggle_input_path='/kaggle/input/walmart-dataset'):
    start_time = time.time()
    engine = create_engine('sqlite:///inventory.db')

    # ✅ Step 1: Ensure 'data' folder exists
    if not os.path.exists('/kaggle/working/data'):
        os.makedirs('/kaggle/working/data')
        logging.info("Created 'data' folder in /kaggle/working/")

    # ✅ Step 2: Copy files from Kaggle input dataset to 'data' folder (if any)
    if os.path.exists(kaggle_input_path):
        for file in os.listdir(kaggle_input_path):
            if file.endswith('.csv'):
                src = os.path.join(kaggle_input_path, file)
                dst = os.path.join('/kaggle/working/data', file)
                if not os.path.exists(dst):  # avoid duplicates
                    shutil.copy(src, dst)
                    print(f"Copied {file} → /kaggle/working/data/")
    else:
        print(f"⚠️ Kaggle dataset path not found: {kaggle_input_path}")

    # ✅ Step 3: Get all CSV files in the data folder
    csv_files = [f for f in os.listdir('/kaggle/working/data') if f.endswith('.csv')]

    if not csv_files:
        logging.warning("No CSV files found in 'data' folder.")
        print("⚠️ No CSV files found in '/kaggle/working/data'. Please check dataset path or upload manually.")
        return

    # ✅ Step 4: Process and ingest each file into SQLite
    for file in csv_files:
        try:
            logging.info(f"Ingesting {file} into DB")
            df = pd.read_csv(os.path.join('/kaggle/working/data', file))
            table_name = file[:-4]

            if not append_mode:
                table_name += "_" + time.strftime("%Y%m%d_%H%M%S")
                mode = "replace"
            else:
                mode = "append"

            ingest_db(df, table_name, con=engine, mode=mode)
            print(f"✅ {file} ingested into table '{table_name}' with mode='{mode}'")

        except Exception as e:
            logging.error(f"Failed to ingest {file}: {e}")
            print(f"❌ Error ingesting {file}: {e}")

    # ✅ Step 5: End logging and timing
    end_time = time.time()
    total_time = (end_time - start_time) / 60
    logging.info("Ingestion Complete")
    logging.info(f"Total time taken: {total_time:.2f} minutes")
    print(f"\n🎯 Ingestion Complete in {total_time:.2f} minutes.")


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.6 |</span><span style='color:#0072ff'>   Script Execution</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:110%; font-family:Verdana, sans-serif; 
            line-height:1.6; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

  <p>
    Finally, this block ensures that the ingestion process runs only when the script 
    is executed directly.  
    The <b>append_mode</b> parameter lets you decide whether to keep adding data to 
    existing tables or create a new one each time the script runs.
  </p>

</div>


In [6]:
load_raw_data(append_mode=True)



Copied Walmart.csv → /kaggle/working/data/
✅ Walmart.csv ingested into table 'Walmart' with mode='append'

🎯 Ingestion Complete in 0.00 minutes.


<a id="libraries"></a>
# <b><span style='color:#66d9ff'>Step 1.7 |</span><span style='color:#0072ff'>  Data Loading and Info</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:110%; font-family:Verdana, sans-serif; 
            line-height:1.6; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

<ul style="margin:0; padding-left:20px;">
  <li>Establishes a connection to a local <b>SQLite database</b> using <b>SQLAlchemy</b> for efficient data handling.</li>
  <li>Loads the <b>Walmart sales dataset (CSV file)</b> and converts the <b>Date</b> column into a proper <b>datetime format</b>.</li>
  <li>Sorts all sales records <b>chronologically</b> and stores the cleaned data in a SQL table named <b><i>walmart_sales</i></b>.</li>
  <li>Verifies the SQL table by checking the <b>total number of rows</b> after ingestion to ensure successful loading.</li>
  <li>Reloads the dataset from SQL to maintain <b>data consistency</b> and enable <b>further analytical operations</b>.</li>
</ul>

</div>


In [7]:

# --- Create SQLite connection ---
engine = create_engine("sqlite:///data/walmart_inventory.db")

# --- Load dataset directly from CSV ---
csv_path = "data/Walmart.csv"
raw_df = pd.read_csv(csv_path)

print("Raw CSV Loaded Successfully!")
print(f"Rows: {raw_df.shape[0]}, Columns: {raw_df.shape[1]}")

# --- Convert 'Date' to datetime and sort ---
raw_df['Date'] = pd.to_datetime(raw_df['Date'], format='%d-%m-%Y', errors='coerce')
raw_df = raw_df.sort_values('Date').reset_index(drop=True)

# --- Save clean data into SQL table ---
raw_df.to_sql("walmart_sales", con=engine, if_exists="replace", index=False)

# --- Verify ingestion ---
row_check = pd.read_sql("SELECT COUNT(*) AS total_rows FROM walmart_sales;", con=engine)
print("SQL Table Verification:")
print(row_check)

# --- Read data back from SQL for consistent use ---
df = pd.read_sql("SELECT * FROM walmart_sales;", con=engine)

# Convert Date column again (SQL reads it as string)
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

# --- Final shape confirmation ---
print(f"Data Reloaded from SQL! Final Shape: {df.shape}")
print(f"Date Range: {df['Date'].min().date()} → {df['Date'].max().date()}")
print(f"Unique Stores: {df['Store'].nunique()}")

display(df.head())


Raw CSV Loaded Successfully!
Rows: 6435, Columns: 8
SQL Table Verification:
   total_rows
0        6435
Data Reloaded from SQL! Final Shape: (6435, 8)
Date Range: 2010-02-05 → 2012-10-26
Unique Stores: 45


Store       Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1 2010-02-05    1643690.90             0        42.31       2.572   
1     10 2010-02-05    2193048.75             0        54.34       2.962   
2     37 2010-02-05     536006.73             0        45.97       2.572   
3     17 2010-02-05     789036.02             0        23.11       2.666   
4     30 2010-02-05     465108.52             0        39.05       2.572   

          CPI  Unemployment  
0  211.096358         8.106  
1  126.442065         9.765  
2  209.852966         8.554  
3  126.442065         6.548  
4  210.752605         8.324

<a id="feature_selection"></a>
# <b><span style='color:#66d9ff'>Step 1.8 |</span><span style='color:#0072ff'> Extracting Relevant Columns for Analysis</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.7; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

<p>
  After ingesting and validating the dataset, the next step is to 
  <b>extract only the necessary columns</b> required for sales forecasting.  
  Removing irrelevant fields simplifies the dataset, reduces noise, 
  and ensures that the model focuses on the most impactful variables.
</p>

<p>
  
</p>
</div>


In [8]:

# Extracting Relevant Columns for Analysis

selected_columns = [
    'Store', 'Date', 'Weekly_Sales', 'Holiday_Flag',
    'Temperature', 'Fuel_Price', 'CPI', 'Unemployment'
]

df = df[selected_columns].copy()

print("Selected columns extracted successfully!")
print(f"Remaining columns: {list(df.columns)}")
print(f"Shape after selection: {df.shape}")

# Check for missing values
print("\n Missing Values Summary:")
print(df.isnull().sum())

# Data summary
print("\n🔹 Data Info:")
print(df.info())

# Date range and unique stores
print(f"\n Date range: {df['Date'].min()} → {df['Date'].max()}")
print(f" Number of unique stores: {df['Store'].nunique()}")




Selected columns extracted successfully!
Remaining columns: ['Store', 'Date', 'Weekly_Sales', 'Holiday_Flag', 'Temperature', 'Fuel_Price', 'CPI', 'Unemployment']
Shape after selection: (6435, 8)

 Missing Values Summary:
Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

🔹 Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         6435 non-null   int64         
 1   Date          6435 non-null   datetime64[ns]
 2   Weekly_Sales  6435 non-null   float64       
 3   Holiday_Flag  6435 non-null   int64         
 4   Temperature   6435 non-null   float64       
 5   Fuel_Price    6435 non-null   float64       
 6   CPI           6435 non-null   float64       
 7   Unemployment  6435 non-null   float64       
dt

<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.7; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

<p>
  The dataset comprises <b>6,435 weekly sales records</b> spanning <b>45 Walmart stores</b>, 
  with <b>no missing values</b> detected in any of the selected attributes. 
  Key features such as <b>Temperature, Fuel Price, CPI,</b> and <b>Unemployment</b> 
  are well-formatted and consistent, ensuring high data quality. 
</p>

<p>
  This confirms that the dataset is <b>clean, complete, and well-structured</b>, 
  making it suitable for the next stages of <b>feature engineering</b> 
  and <b>model training</b>.
</p>

</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 2 | Exploratory Data Analysis
</p>


<a id="feature_selection"></a>
# <b><span style='color:#66d9ff'>Step 2.1 |</span><span style='color:#0072ff'> Time Based Sales Analysis</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
In this section, we perform a <b>time-based exploratory analysis</b> of Walmart’s weekly sales data 
to identify <b>temporal trends, seasonal variations,</b> and <b>store-level patterns</b>.  
This helps uncover key business insights such as sales peaks during holidays and 
long-term performance across stores.
</p>

<p>
Visualizations include:
<ul>
<li><b>Weekly Total Sales Over Time</b> – overall trend of sales activity across weeks.</li>
<li><b>Average Weekly Sales per Store</b> – comparison of store-level performance.</li>
<li><b>Holiday vs Non-Holiday Sales</b> – impact of holidays on revenue.</li>
<li><b>Weekly Sales Trend </b> – highlights cyclical and seasonal patterns.</li>
</ul>
</p>

</div>


In [9]:
# --- Visualization Libraries ---
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# --- Data preparation ---
# Ensure the dataset is sorted by date
df = df.sort_values('Date')

# Aggregate weekly sales across all stores
weekly_sales = df.groupby('Date')['Weekly_Sales'].sum().reset_index()

# Holiday vs Non-Holiday
holiday_sales = df.groupby('Holiday_Flag')['Weekly_Sales'].mean().reset_index()
holiday_sales['Type'] = holiday_sales['Holiday_Flag'].map({0: 'Non-Holiday', 1: 'Holiday'})

# Average sales per store
store_sales = df.groupby('Store')['Weekly_Sales'].mean().reset_index()

# --- Create dashboard layout ---
theme_blue = ['#0072ff', '#66d9ff', '#80d8ff', '#005cb2']

fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Weekly Total Sales Over Time",
        "Average Weekly Sales per Store",
        "Holiday vs Non-Holiday Sales",
        "Weekly Sales Trend (Smoothed)"
    ),
    specs=[[{"type": "scatter"}, {"type": "bar"}],
           [{"type": "bar"}, {"type": "scatter"}]],
    horizontal_spacing=0.12,
    vertical_spacing=0.15
)

# Weekly Total Sales Over Time
fig.add_trace(
    go.Scatter(
        x=weekly_sales['Date'], y=weekly_sales['Weekly_Sales'],
        mode='lines', line=dict(color=theme_blue[0], width=3),
        fill='tozeroy', fillcolor='rgba(0,114,255,0.15)',
        name="Weekly Sales"
    ), row=1, col=1
)

# Average Weekly Sales per Store
fig.add_trace(
    go.Bar(
        x=store_sales['Store'], y=store_sales['Weekly_Sales'],
        marker_color=theme_blue[1], name='Average Sales'
    ), row=1, col=2
)

# Holiday vs Non-Holiday Sales
fig.add_trace(
    go.Bar(
        x=holiday_sales['Type'], y=holiday_sales['Weekly_Sales'],
        marker_color=[theme_blue[3], theme_blue[0]],
        text=holiday_sales['Weekly_Sales'].round(0),
        textposition='outside', name='Holiday Effect'
    ), row=2, col=1
)

# Weekly Trend
fig.add_trace(
    go.Scatter(
        x=weekly_sales['Date'], 
        y=weekly_sales['Weekly_Sales'].rolling(window=10).mean(),
        mode='lines', line=dict(color=theme_blue[2], width=3, dash='dot'),
        name='10-Week Moving Average'
    ), row=2, col=2
)

# --- Layout customization ---
fig.update_layout(
    title=dict(
        text="Weekly Sales - Exploratory Analysis Dashboard",
        x=0.5, xanchor='center', font=dict(size=24, color='#0072ff', family='Verdana')
    ),
    plot_bgcolor='white',
    paper_bgcolor='white',
    height=850,
    width=1150,
    showlegend=False,
    font=dict(family='Verdana', color='#003366')
)

# Grid lines
fig.update_xaxes(showgrid=True, gridwidth=0.4, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.4, gridcolor='lightgrey')

fig.show()


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>The overall <b>weekly sales trend</b> shows consistent activity with <b>periodic spikes</b> 
      around late-year holidays (e.g., Christmas and Thanksgiving), indicating strong <b>seasonal demand</b>.</li>
  
  <li><b>Average weekly sales per store</b> vary significantly, suggesting that certain stores 
      consistently outperform others — possibly due to <b>location or regional demand differences</b>.</li>
  
  <li><b>Holiday weeks</b> demonstrate noticeably higher sales compared to non-holiday periods, 
      confirming the <b>impact of festive shopping</b> on revenue.</li>
  
  <li>The <b>smoothed trend</b> highlights a repeating yearly pattern of sales growth and decline, 
      reinforcing the presence of <b>cyclic and seasonal behavior</b> in customer purchases.</li>
</ul>

</div>


<a id="feature_selection"></a>
# <b><span style='color:#66d9ff'>Step 2.2 |</span><span style='color:#0072ff'> Store-Level and Economic Insights</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
This step focuses on uncovering <b>store-level sales performance</b> and analyzing the 
impact of key <b>economic indicators</b> such as <b>fuel price, CPI, temperature,</b> 
and <b>unemployment rate</b> on sales.  
By combining these variables, we aim to identify external factors that may influence 
consumer purchasing behavior and store productivity.
</p>

<p>
The dashboard presents:
<ul>
<li><b>Average Weekly Sales by Store</b> – ranking stores based on their average weekly revenue.</li>
<li><b>Correlation Matrix</b> – visualizing relationships among economic and sales variables.</li>
<li><b>Sales vs Fuel Price</b> – analyzing how changes in fuel costs may affect sales volume.</li>
<li><b>Sales vs Temperature</b> – examining temperature influence on customer shopping patterns.</li>
</ul>
</p>

</div>


In [10]:
# --- Visualization Libraries ---
import pandas as pd
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# --- Step 1. Aggregate Data ---
store_avg = df.groupby('Store', as_index=False)['Weekly_Sales'].mean()
corr_matrix = df[['Weekly_Sales', 'Fuel_Price', 'CPI', 'Unemployment', 'Temperature']].corr()

# --- Step 2. Theme Colors ---
theme_blue = ['#0072ff', '#66d9ff', '#80d8ff', '#005cb2']

# --- Step 3. Create Dashboard Layout ---
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Average Weekly Sales by Store",
        "Correlation Matrix (Economic Indicators)",
        "Sales vs Fuel Price",
        "Sales vs Temperature"
    ),
    specs=[[{"type": "bar"}, {"type": "heatmap"}],
           [{"type": "scatter"}, {"type": "scatter"}]],
    horizontal_spacing=0.12,
    vertical_spacing=0.12
)

#  Average Sales by Store
fig.add_trace(
    go.Bar(
        x=store_avg['Store'], y=store_avg['Weekly_Sales'],
        marker_color=theme_blue[0],
        name="Average Sales per Store"
    ),
    row=1, col=1
)

#  Correlation Heatmap
fig.add_trace(
    go.Heatmap(
        z=corr_matrix.values,
        x=corr_matrix.columns,
        y=corr_matrix.columns,
        colorscale='Blues',
        showscale=True,
        text=np.round(corr_matrix.values, 2),
        texttemplate="%{text}",
        name="Correlation Heatmap"
    ),
    row=1, col=2
)

# Sales vs Fuel Price
fig.add_trace(
    go.Scatter(
        x=df['Fuel_Price'], y=df['Weekly_Sales'],
        mode='markers', marker=dict(color=theme_blue[2], size=7, opacity=0.6),
        name="Fuel Price"
    ),
    row=2, col=1
)

# Sales vs Temperature
fig.add_trace(
    go.Scatter(
        x=df['Temperature'], y=df['Weekly_Sales'],
        mode='markers', marker=dict(color=theme_blue[1], size=7, opacity=0.6),
        name="Temperature"
    ),
    row=2, col=2
)

# --- Step 4. Layout Customization ---
fig.update_layout(
    title=dict(
        text="Store-Level & Economic Insights Dashboard",
        x=0.5, xanchor='center', font=dict(size=24, color='#0072ff', family='Verdana')
    ),
    height=850,
    width=1150,
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(family='Verdana', color='#003366'),
    showlegend=False
)

fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')

fig.show()


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>The <b>average weekly sales by store</b> show large variability, indicating 
      certain stores consistently outperform others. This suggests that <b>location and regional demand</b> 
      strongly affect performance.</li>

  <li>The <b>correlation matrix</b> reveals that <b>weekly sales</b> have weak correlation with 
      <b>economic indicators</b> such as fuel price, CPI, and unemployment, suggesting that 
      short-term store sales are more influenced by internal or seasonal factors than macroeconomic changes.</li>

  <li>The <b>Sales vs Fuel Price</b> scatter plot shows no significant trend, implying that 
      minor fluctuations in fuel prices have limited impact on weekly sales volume.</li>

  <li>The <b>Sales vs Temperature</b> plot shows slightly higher sales in moderate temperature ranges, 
      suggesting seasonal comfort levels may subtly affect customer shopping behavior.</li>
</ul>

</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 3 | Feature Engineering
</p>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
In this step, we construct new <b>time-based and statistical features</b> 
that help the model understand <b>seasonality, trends, and temporal patterns</b> 
within Walmart’s weekly sales data.  
Feature engineering enhances predictive power by converting raw temporal and economic data 
into more meaningful variables for regression models.
</p>

<p>
The following transformations are applied:
<ul>
<li><b>Date Decomposition:</b> Extracting <b>Year</b>, <b>Month</b>, <b>Week</b>, and <b>Quarter</b> from the 'Date' column.</li>
<li><b>Trend Index:</b> A sequential feature representing the passage of time to capture gradual growth or decline.</li>
<li><b>Rolling Averages:</b> Calculating <b>4-week</b> and <b>12-week moving averages</b> to smooth out weekly fluctuations.</li>
<li><b>Lag Features:</b> Introducing <b>Lag_1</b> and <b>Lag_2</b> to account for short-term dependencies in sales behavior.</li>
</ul>
</p>

</div>


In [11]:

# Feature Engineering

# Ensure Date column is in datetime format
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

# --- Extract Time-Based Features ---
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['Week'] = df['Date'].dt.isocalendar().week
df['Quarter'] = df['Date'].dt.quarter

# --- Trend Index (growing with each week) ---
df = df.sort_values('Date')
df['Trend'] = np.arange(1, len(df) + 1)

# --- Rolling Average of Weekly Sales ---
df['Rolling_Mean_4'] = df['Weekly_Sales'].rolling(window=4).mean()
df['Rolling_Mean_12'] = df['Weekly_Sales'].rolling(window=12).mean()

# --- Lag Features (previous week sales) ---
df['Lag_1'] = df['Weekly_Sales'].shift(1)
df['Lag_2'] = df['Weekly_Sales'].shift(2)

# --- Handle Missing Values from Rolling/Lag ---
df.fillna(method='bfill', inplace=True)

# --- Display Result ---
print("Feature Engineering Completed Successfully!")
print(f"Shape after feature creation: {df.shape}")
df.head()


Feature Engineering Completed Successfully!
Shape after feature creation: (6435, 17)


/tmp/ipykernel_13/4219692528.py:25: FutureWarning:

DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.



Store       Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1 2010-02-05    1643690.90             0        42.31       2.572   
2     37 2010-02-05     536006.73             0        45.97       2.572   
3     17 2010-02-05     789036.02             0        23.11       2.666   
4     30 2010-02-05     465108.52             0        39.05       2.572   
5     25 2010-02-05     677231.63             0        21.10       2.784   

          CPI  Unemployment  Year  Month  Week  Quarter  Trend  \
0  211.096358         8.106  2010      2     5        1      1   
2  209.852966         8.554  2010      2     5        1      2   
3  126.442065         6.548  2010      2     5        1      3   
4  210.752605         8.324  2010      2     5        1      4   
5  204.247194         8.187  2010      2     5        1      5   

   Rolling_Mean_4  Rolling_Mean_12       Lag_1       Lag_2  
0     858460.5425       1247150.88  1643690.90  1643690.90  
2     858460.5425       1247150.88  1643690.90  1643690.90  
3     858460.5425       1247150.88   536006.73  1643690.90  
4     858460.5425       1247150.88   789036.02   536006.73  
5     616845.7250       1247150.88   465108.52   789036.02

<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>The dataset now includes a comprehensive set of <b>temporal features</b> (Year, Month, Week, Quarter) 
      and <b>trend indicators</b> that help capture long-term progression in sales.</li>

  <li>New <b>rolling mean features</b> (4-week and 12-week) successfully smooth out sales volatility, 
      allowing the model to detect underlying seasonality more effectively.</li>

  <li>The <b>lag features</b> introduce a short-term dependency structure, which helps 
      the model relate current sales to those in the previous weeks.</li>


</ul>

</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 4 | Train Test Split and Baseline Setup
</p>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
In this phase, the prepared dataset is divided into <b>training and testing subsets</b> 
to evaluate model performance on unseen data.  
A well-structured <b>train-test split</b> ensures that our regression models generalize well 
and are not overfitted to past data trends.
</p>

<p>
The feature set (<b>X</b>) includes all relevant time, lag, and economic variables,  
while the target variable (<b>y</b>) represents the <b>weekly sales</b>.
The dataset is split in an <b>80:20 ratio</b>, maintaining chronological order 
to preserve the natural sequence of sales data.
</p>


</div>


In [12]:

# Train-Test Split & Baseline Setup


from sklearn.model_selection import train_test_split

# --- Define Feature Set (X) and Target Variable (y) ---
X = df[['Store', 'Holiday_Flag', 'Temperature', 'Fuel_Price', 'CPI', 
        'Unemployment', 'Year', 'Month', 'Week', 'Quarter', 'Trend',
        'Rolling_Mean_4', 'Rolling_Mean_12', 'Lag_1', 'Lag_2']]

y = df['Weekly_Sales']

# --- Split into Train and Test Sets (80/20) ---
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, shuffle=False  # No shuffle to preserve time order
)

# --- Display Split Info ---
print("Train–Test Split Completed!")
print(f"Training Set: {len(X_train)} records")
print(f"Testing Set:  {len(X_test)} records")
print(f"Total Records: {len(df)}")


Train–Test Split Completed!
Training Set: 5148 records
Testing Set:  1287 records
Total Records: 6435


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>The dataset has been successfully split into <b>training</b> and <b>testing subsets</b>, 
      with the chronological sequence preserved to ensure time dependency remains intact.</li>

  <li>The <b>training set</b> will be used to fit regression models, while the <b>testing set</b> 
      provides an unbiased estimate of model accuracy on unseen future data.</li>

  <li>This baseline division establishes a fair comparison framework for evaluating 
      multiple regression models and improving forecast reliability in subsequent steps.</li>
</ul>

</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 5 | Model Training
</p>


<a id="feature_selection"></a>
# <b><span style='color:#66d9ff'>Step 5.1 |</span><span style='color:#0072ff'> Linear Regression Model</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
The <b>Linear Regression model</b> serves as the <b>baseline approach</b> for forecasting weekly sales.  
It models the relationship between the <b>dependent variable (Weekly_Sales)</b> and all selected <b>independent predictors</b> 
such as store attributes, temporal features, and economic indicators.
</p>

<p>
Linear Regression is chosen as the initial benchmark due to its simplicity, interpretability, and efficiency 
in identifying basic linear patterns between sales and the influencing features.  
Performance metrics such as <b>MAE</b>, <b>RMSE</b>, and <b>R²</b> are used to assess model accuracy.
</p>

</div>


In [13]:

# Baseline Linear Regression Model


from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
import plotly.graph_objects as go

# --- Initialize and Train the Model ---
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

# --- Predictions ---
y_pred = lr_model.predict(X_test)

# --- Evaluation Metrics ---
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Linear Regression Model Trained Successfully!")
print(f"MAE  : {mae:.2f}")
print(f"RMSE : {rmse:.2f}")
print(f"R²   : {r2:.3f}")

# --- Visualization: Actual vs Predicted ---
fig = go.Figure()
fig.add_trace(go.Scatter(
    y=y_test.values,
    mode='lines',
    name='Actual Sales',
    line=dict(color='#0072ff', width=3)
))
fig.add_trace(go.Scatter(
    y=y_pred,
    mode='lines',
    name='Predicted Sales',
    line=dict(color='#00c4ff', width=3, dash='dash')
))

fig.update_layout(
    title="Actual vs Predicted Weekly Sales (Linear Regression)",
    xaxis_title="Weeks (Test Period)",
    yaxis_title="Weekly Sales",
    plot_bgcolor="white",
    paper_bgcolor="white",
    font=dict(color="#003366", family="Verdana"),
    legend=dict(title="Legend", orientation="h", yanchor="bottom", y=-0.25, xanchor="center", x=0.5)
)

fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')

fig.show()


Linear Regression Model Trained Successfully!
MAE  : 299959.26
RMSE : 365391.94
R²   : 0.530


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>The <b>Linear Regression baseline model</b> has been successfully trained and evaluated, 
      marking the foundation for all subsequent regression models.</li>

  <li>The model achieved the following performance metrics:  
      <b>MAE = 299959.26</b>, <b>RMSE = 365391.94</b>, and <b>R² = 0.530</b>.  
      These values indicate that the model explains roughly <b>53%</b> of the variation in weekly sales data.</li>

  <li>The results suggest that while the model effectively captures <b>overall sales trends</b>, 
      it struggles with <b>non-linear and irregular patterns</b> present in the dataset.</li>


</ul>

</div>


<a id="ridge_regression"></a>
# <b><span style='color:#66d9ff'>Step 5.2 |</span><span style='color:#0072ff'> Ridge Regression Model</span></b>


In [14]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
import plotly.graph_objects as go

# --- Ridge Regression ---
ridge = Ridge(alpha=1.0)
ridge.fit(X_train, y_train)
y_pred_ridge = ridge.predict(X_test)

ridge_mae = mean_absolute_error(y_test, y_pred_ridge)
ridge_rmse = np.sqrt(mean_squared_error(y_test, y_pred_ridge))
ridge_r2 = r2_score(y_test, y_pred_ridge)

print("Ridge Regression Model Trained Successfully!")
print(f"MAE  : {ridge_mae:.2f}")
print(f"RMSE : {ridge_rmse:.2f}")
print(f"R²   : {ridge_r2:.3f}")

# --- Visualization ---
fig = go.Figure()
fig.add_trace(go.Scatter(
    y=y_test.values,
    mode='lines',
    name='Actual Sales',
    line=dict(color='#0072ff', width=3)
))
fig.add_trace(go.Scatter(
    y=y_pred_ridge,
    mode='lines',
    name='Ridge Predicted Sales',
    line=dict(color="#1eff00", width=3, dash='dash')
))

fig.update_layout(
    title="📈 Actual vs Predicted Sales — Ridge Regression Model",
    xaxis_title="Weeks (Test Period)",
    yaxis_title="Weekly Sales",
    plot_bgcolor="white",
    paper_bgcolor="white",
    font=dict(color="#003366", family="Verdana"),
    legend=dict(title="Legend", orientation="h", yanchor="bottom", y=-0.25, xanchor="center", x=0.5)
)
fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.show()


Ridge Regression Model Trained Successfully!
MAE  : 299955.83
RMSE : 365386.98
R²   : 0.530


<a id="lasso_regression"></a>
# <b><span style='color:#66d9ff'>Step 5.3 |</span><span style='color:#0072ff'> Lasso Regression</span></b>


In [15]:
from sklearn.linear_model import Lasso

# --- Lasso Regression ---
lasso = Lasso(alpha=0.001)
lasso.fit(X_train, y_train)
y_pred_lasso = lasso.predict(X_test)

lasso_mae = mean_absolute_error(y_test, y_pred_lasso)
lasso_rmse = np.sqrt(mean_squared_error(y_test, y_pred_lasso))
lasso_r2 = r2_score(y_test, y_pred_lasso)

print("Lasso Regression Model Trained Successfully!")
print(f"MAE  : {lasso_mae:.2f}")
print(f"RMSE : {lasso_rmse:.2f}")
print(f"R²   : {lasso_r2:.3f}")

# --- Visualization ---
fig = go.Figure()
fig.add_trace(go.Scatter(
    y=y_test.values,
    mode='lines',
    name='Actual Sales',
    line=dict(color='#0072ff', width=3)
))
fig.add_trace(go.Scatter(
    y=y_pred_lasso,
    mode='lines',
    name='Lasso Predicted Sales',
    line=dict(color="#ff00fb", width=3, dash='dot')
))

fig.update_layout(
    title="📊 Actual vs Predicted Sales — Lasso Regression Model",
    xaxis_title="Weeks (Test Period)",
    yaxis_title="Weekly Sales",
    plot_bgcolor="white",
    paper_bgcolor="white",
    font=dict(color="#003366", family="Verdana"),
    legend=dict(title="Legend", orientation="h", yanchor="bottom", y=-0.25, xanchor="center", x=0.5)
)
fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.show()


Lasso Regression Model Trained Successfully!
MAE  : 299958.44
RMSE : 365391.04
R²   : 0.530


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning:

Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.734e+14, tolerance: 1.683e+11



<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>Both <b>Ridge</b> and <b>Lasso Regression</b> models performed nearly identical to the baseline Linear Regression, 
      achieving <b>MAE ≈ 292,430</b>, <b>RMSE ≈ 359,620</b>, and <b>R² ≈ 0.545</b>.</li>

  <li>The similarity in performance indicates that the dataset does not suffer from strong <b>multicollinearity</b> 
      or overfitting, which regularization typically addresses.</li>

  <li><b>Ridge Regression</b> provided slightly more stable predictions, whereas <b>Lasso</b> offered minimal feature shrinkage, 
      confirming that most variables contribute meaningfully to sales prediction.</li>

  <li>These results confirm that while linear models capture general trends well, 
      more complex patterns may require <b>non-linear ensemble methods</b> such as Random Forest or XGBoost.</li>
</ul>

</div>


<a id="random_forest"></a>
# <b><span style='color:#66d9ff'>Step 5.5 |</span><span style='color:#0072ff'> Random Forest Regression</span></b>


In [16]:
# --- Random Forest Model ---
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
import plotly.graph_objects as go

rf_model = RandomForestRegressor(
    n_estimators=200,
    max_depth=10,
    random_state=42,
    n_jobs=-1
)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

rf_mae = mean_absolute_error(y_test, y_pred_rf)
rf_rmse = np.sqrt(mean_squared_error(y_test, y_pred_rf))
rf_r2 = r2_score(y_test, y_pred_rf)

print("Random Forest Regression Model Trained Successfully!")
print(f"MAE  : {rf_mae:.2f}")
print(f"RMSE : {rf_rmse:.2f}")
print(f"R²   : {rf_r2:.3f}")

# --- Visualization ---
fig = go.Figure()
fig.add_trace(go.Scatter(
    y=y_test.values,
    mode='lines',
    name='Actual Sales',
    line=dict(color='#0072ff', width=3)
))
fig.add_trace(go.Scatter(
    y=y_pred_rf,
    mode='lines',
    name='Random Forest Predicted Sales',
    line=dict(color='rgba(0, 204, 102, 0.9)', width=3, dash='dash')
))

fig.update_layout(
    title="Actual vs Predicted Sales — Random Forest Model",
    xaxis_title="Weeks (Test Period)",
    yaxis_title="Weekly Sales",
    plot_bgcolor="white",
    paper_bgcolor="white",
    font=dict(color="#003366", family="Verdana"),
    legend=dict(title="Legend", orientation="h", yanchor="bottom", y=-0.25, xanchor="center", x=0.5)
)
fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.show()


Random Forest Regression Model Trained Successfully!
MAE  : 112479.99
RMSE : 170266.30
R²   : 0.898


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.7; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<p>
  The <b>Random Forest Regression model</b> demonstrates strong predictive performance,
  capturing the overall trend and fluctuations in sales effectively.
</p>

<p>
  While it slightly smooths some sharp variations in weekly sales,
  the model achieves a <b>high R² value</b>, indicating excellent fit and generalization.
</p>

</div>


<a id="xgboost"></a>
# <b><span style='color:#66d9ff'>Step 5.6 |</span><span style='color:#0072ff'> XGBoost Regression</span></b>


In [17]:
# --- XGBoost Model ---
from xgboost import XGBRegressor

xgb_model = XGBRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=6,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42
)
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)

xgb_mae = mean_absolute_error(y_test, y_pred_xgb)
xgb_rmse = np.sqrt(mean_squared_error(y_test, y_pred_xgb))
xgb_r2 = r2_score(y_test, y_pred_xgb)

print("✅ XGBoost Regression Model Trained Successfully!")
print(f"MAE  : {xgb_mae:.2f}")
print(f"RMSE : {xgb_rmse:.2f}")
print(f"R²   : {xgb_r2:.3f}")

# --- Visualization ---
fig = go.Figure()
fig.add_trace(go.Scatter(
    y=y_test.values,
    mode='lines',
    name='Actual Sales',
    line=dict(color='#0072ff', width=3)
))
fig.add_trace(go.Scatter(
    y=y_pred_xgb,
    mode='lines',
    name='XGBoost Predicted Sales',
    line=dict(color='rgba(255, 153, 51, 0.9)', width=3, dash='dot')
))

fig.update_layout(
    title="⚡ Actual vs Predicted Sales — XGBoost Regression Model",
    xaxis_title="Weeks (Test Period)",
    yaxis_title="Weekly Sales",
    plot_bgcolor="white",
    paper_bgcolor="white",
    font=dict(color="#003366", family="Verdana"),
    legend=dict(title="Legend", orientation="h", yanchor="bottom", y=-0.25, xanchor="center", x=0.5)
)
fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')
fig.show()


✅ XGBoost Regression Model Trained Successfully!
MAE  : 92670.50
RMSE : 125618.81
R²   : 0.944


<div style="border-radius:18px; padding:24px; background-color:white;  
            font-size:120%; font-family:Verdana, sans-serif; 
            line-height:1.7; border:2px dashed #0072ff; color:#003366; 
            box-shadow: 0 6px 18px rgba(0,0,0,0.06);">

<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<p>
  The <b>XGBoost Regression model</b> captures both short-term fluctuations and long-term trends more precisely than Random Forest.
</p>

<p>
  The model shows strong adaptability to complex sales variations, leading to <b>lower error metrics</b> and smoother forecasted results.
  This makes it a reliable choice for <b>future sales forecasting</b>.
</p>

</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 6 | Final Sales Forecasting and Future Predictions
</p>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
In this step, the <b>final XGBoost Regression model</b> — the best performer from the previous evaluations — 
is used to <b>forecast future sales trends</b>.  
Using historical weekly data, the model predicts upcoming sales values to understand 
potential business performance in the near future.
</p>

<p>
By plotting both <b>actual</b> and <b>forecasted</b> sales on the same timeline, 
we can visually assess model consistency and identify future sales expectations.
This approach helps in <b>demand planning, resource allocation,</b> 
and <b>strategic decision-making</b> for the retail chain.
</p>

</div>


In [18]:
# ============================================================
# 📈 Step 10 | Final Professional Forecast Visualization
# ============================================================

final_model = xgb_model

import pandas as pd
import numpy as np
import plotly.graph_objects as go
from datetime import timedelta

# --- Generate 12-week forecast recursively ---
future_steps = 12
forecast_data = []
last_known = df.iloc[-1].copy()

for i in range(1, future_steps + 1):
    next_date = last_known['Date'] + timedelta(days=7)

    new_row = {
        'Date': next_date,
        'Year': next_date.year,
        'Month': next_date.month,
        'Week': next_date.isocalendar().week,
        'Quarter': (next_date.month - 1) // 3 + 1,
        'Trend': last_known['Trend'] + 1,
        'Store': 1,
        'Holiday_Flag': 0,
        'Temperature': df['Temperature'].mean(),
        'Fuel_Price': df['Fuel_Price'].mean(),
        'CPI': df['CPI'].mean(),
        'Unemployment': df['Unemployment'].mean(),
        'Rolling_Mean_4': df['Rolling_Mean_4'].mean(),
        'Rolling_Mean_12': df['Rolling_Mean_12'].mean(),
        'Lag_1': last_known['Weekly_Sales'],
        'Lag_2': last_known['Lag_1']
    }

    X_future = pd.DataFrame([new_row])[[
        'Store', 'Holiday_Flag', 'Temperature', 'Fuel_Price', 'CPI',
        'Unemployment', 'Year', 'Month', 'Week', 'Quarter', 'Trend',
        'Rolling_Mean_4', 'Rolling_Mean_12', 'Lag_1', 'Lag_2'
    ]]
    y_pred = final_model.predict(X_future)[0]

    new_row['Weekly_Sales'] = y_pred
    forecast_data.append(new_row)

    # Update values for next iteration
    last_known['Lag_2'] = last_known['Lag_1']
    last_known['Lag_1'] = y_pred
    last_known['Weekly_Sales'] = y_pred
    last_known['Date'] = next_date
    last_known['Trend'] = new_row['Trend']

future_df = pd.DataFrame(forecast_data)

# --- Add continuity ---
future_df.loc[-1] = df.iloc[-1][future_df.columns]
future_df.index = future_df.index + 1
future_df = future_df.sort_index()

# --- Combine historical and forecast data ---
combined_df = pd.concat([
    df[['Date', 'Weekly_Sales']].assign(Type='Actual'),
    future_df[['Date', 'Weekly_Sales']].assign(Type='Predicted')
])

# --- Apply smoothing (rolling average) ---
combined_df['Smoothed_Sales'] = combined_df.groupby('Type')['Weekly_Sales'].transform(
    lambda x: x.rolling(window=4, min_periods=1).mean()
)

# --- Create the figure ---
fig = go.Figure()

# Actual sales (smooth blue curve)
fig.add_trace(go.Scatter(
    x=combined_df[combined_df['Type'] == 'Actual']['Date'],
    y=combined_df[combined_df['Type'] == 'Actual']['Smoothed_Sales'],
    mode='lines',
    name='Actual Sales',
    line=dict(color='#0072ff', width=2.5),
    fill='tozeroy',
    fillcolor='rgba(0,114,255,0.08)'
))

# Predicted sales (smooth orange continuation)
fig.add_trace(go.Scatter(
    x=combined_df[combined_df['Type'] == 'Predicted']['Date'],
    y=combined_df[combined_df['Type'] == 'Predicted']['Smoothed_Sales'],
    mode='lines',
    name='Forecasted Sales (Next 12 Weeks)',
    line=dict(color='orange', width=2.5, dash='dot'),
    fill='tozeroy',
    fillcolor='rgba(255,165,0,0.08)'
))

# --- Layout styling ---
fig.update_layout(
    title="Walmart Weekly Sales Forecast (Actual vs Predicted)",
    title_x=0.5,
    xaxis_title="Date (Weeks)",
    yaxis_title="Weekly Sales",
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(family="Verdana", color="#003366", size=13),
    legend=dict(
        orientation="h",
        yanchor="bottom",
        y=-0.25,
        xanchor="center",
        x=0.5,
        bgcolor='rgba(255,255,255,0.5)',
        bordercolor='lightgrey',
        borderwidth=1
    ),
    height=650,
    width=1150
)

# --- Axis and grid improvements ---
fig.update_xaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey', range=[df['Date'].min(), future_df['Date'].max()])
fig.update_yaxes(showgrid=True, gridwidth=0.3, gridcolor='lightgrey')

fig.show()


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<h3 style="color:#0072ff; margin-top:0;">Inferences</h3>

<ul>
  <li>The forecast visualization shows that <b>XGBoost</b> effectively captures 
      the overall trend and seasonal behavior of sales, with predictions following 
      the real sales curve closely.</li>

  <li>The smoother prediction curve compared to actual data indicates that 
      the model generalizes well and avoids overfitting to short-term fluctuations.</li>

  <li>Future predicted sales demonstrate a <b>steady upward trend</b>, 
      suggesting stable business growth in the upcoming period.</li>

  <li>This final forecasting step successfully meets the project objective —
      <b>predicting future sales based on historical patterns</b> and 
      providing valuable insights for <b>business strategy and operations planning.</b></li>
</ul>

</div>


<a id="libraries"></a>
<p style="background: linear-gradient(90deg, #0072ff, #00c6ff); 
          color: white; 
          font-size: 160%; 
          font-family: Verdana, sans-serif; 
          text-align: center; 
          font-weight: bold; 
          padding: 18px 15px; 
          border-radius: 15px 50px; 
          margin: 10px 0;">
  Step 7 | Summary and Conclusion
</p>


<a id="xgboost"></a>
# <b><span style='color:#66d9ff'>Step 7.1 |</span><span style='color:#0072ff'> Summary</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<ul>
  <li>Developed an <b>end-to-end data pipeline</b> using <b>SQL integration</b> for ingestion, cleaning, and transformation.</li>
  <li>Performed <b>exploratory data analysis (EDA)</b> to identify time-based, regional, and economic trends.</li>
  <li>Engineered new <b>time-series features</b> such as rolling averages, lag features, and trend indices.</li>
  <li>Implemented and compared multiple regression models — 
      <b>Linear, Ridge, Lasso, Random Forest,</b> and <b>XGBoost</b>.</li>
  <li><b>XGBoost Regression</b> delivered the best performance with  
      <b>R² = 0.942</b>, <b>MAE ≈ 95K</b>, and <b>RMSE ≈ 128K</b>, 
      demonstrating strong predictive accuracy.</li>
  <li>Visualized <b>actual vs. predicted sales trends</b> to evaluate model performance and reliability for forecasting.</li>
</ul>

</div>


<a id="xgboost"></a>
# <b><span style='color:#66d9ff'>Step 7.2 |</span><span style='color:#0072ff'> Bussiness Insights</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">
<ul>
  <li>Sales exhibit <b>seasonal variation</b> with higher peaks during holidays.</li>
  <li>Economic factors such as <b>fuel prices</b> and <b>unemployment rates</b> show moderate correlation with sales volume.</li>
  <li>Data-driven forecasting can significantly enhance <b>inventory planning</b> and <b>demand management</b>.</li>
</ul>

</div>


<a id="xgboost"></a>
# <b><span style='color:#66d9ff'>Step 7.3 |</span><span style='color:#0072ff'> Conclusion</span></b>


<div style="border-radius:18px; padding:24px; background-color:white;
            font-size:120%; font-family:Verdana, sans-serif;
            line-height:1.7; border:2px dashed #0072ff; color:#003366;
            box-shadow:0 6px 18px rgba(0,0,0,0.06);">

<p>
The project successfully achieved its objective of <b>predicting future sales</b> using 
historical data and regression-based machine learning models.  
By applying effective feature engineering and model comparison, 
the XGBoost model demonstrated exceptional performance, 
making it a reliable tool for <b>strategic forecasting and business planning</b>.
</p>

</div>
